### PR TITLE
[21696] Fix Secure Discovery Server client disposals guid and handshake_handle assertion

### PR DESCRIPTION
**Pull request by Mario-DL**
## Description

This PR addresses two issues regarding `Secure Discovery Server`

* Disposal `guid` was not beaing properly send from clients to server.
* Prevents an assertion triggering in the `SecurityManager` as a consequence of not being able to retrieve a new change in the `participant_stateless_msg_writer`

@Mergifyio backport 2.10.x

## Contributor Checklist

- [X] Commit messages follow the project guidelines. 
- [X] The code follows the style guidelines of this project. 
- [X] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added; the added tests pass locally 
- [X] Any new/modified methods have been properly documented using Doxygen. 
- **N/A** Any new configuration API has an equivalent XML API (with the corresponding XSD extension) 
- [X] Changes are backport compatible: they do **NOT** break ABI nor change library core behavior. 
- [X] Changes are API compatible. 
- **N/A** New feature has been added to the `versions.md` file (if applicable).
- **N/A** New feature has been documented/Current behavior is correctly described in the documentation. 
    
- [X] Applicable backports have been included in the description.

## Reviewer Checklist

- [x] The PR has a milestone assigned.
- [x] The title and description correctly express the PR's purpose.
- [x] Check contributor checklist is correct.
- [ ] If this is a critical bug fix, backports to the critical-only supported branches have been requested.
- [x] Check CI results: changes do not issue any warning.
- [x] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

File: src/cpp/rtps/builtin/discovery/participant/PDPClient.cpp [hunk 1]
```diff
@@ -769,9 +769,17 @@ void PDPClient::announceParticipantState(
                         // if we are matched to a server report demise
                         if (svr.is_connected)
                         {
+                            GuidPrefix_t srv_guid_prefix = svr.guidPrefix;
+
+                            if (getRTPSParticipant()->is_secure())
+                            {
+                                // Need the mangled guid prefix in this case
+                                srv_guid_prefix = get_participant_proxy_data(svr.guidPrefix)->m_guid.guidPrefix;
+                            }
```
Comment (JesusPoderoso) [inline src/cpp/rtps/builtin/discovery/participant/PDPClient.cpp:778]: Based on the CI results, it looks like this should be used with security guards.

File: test/dds/communication/security/CMakeLists.txt [hunk 2]
```diff
@@ -393,6 +393,48 @@ if(SECURITY)
...
+
+        # Set test with label NoMemoryCheck
+        set_property(TEST SecureDiscoverServerMultipleClientsDisposalsReceived PROPERTY LABELS "NoMemoryCheck")
+
+        if(WIN32)
+            string(REPLACE ";" "\\;" WIN_PATH "$ENV{PATH}")
+            set_property(TEST SecureDiscoverServerMultipleClientsDisposalsReceived APPEND PROPERTY ENVIRONMENT
+                "PATH=$<TARGET_FILE_DIR:${PROJECT_NAME}>\\;$<TARGET_FILE_DIR:fastcdr>\\;${WIN_PATH}")
+        endif()
+
```
Comment (JesusPoderoso) [inline test/dds/communication/security/CMakeLists.txt:437]: Fix EOF

File: test/dds/communication/security/multiple_secure_ds_pubsub_secure_crypto_communication.py [hunk 3]
```diff
@@ -152,75 +252,64 @@ def run(args):
...
+    cleanup(pub_procs, sub_procs, ds_procs)
+    terminate(terminate_ok)
 
 if __name__ == '__main__':
 
     # Parse arguments
     args = ParseOptions()
 
-    run(args.args)
+    run(args.args)
```
Comment (JesusPoderoso) [inline test/dds/communication/security/multiple_secure_ds_pubsub_secure_crypto_communication.py:315]: EOF

**Review by JesusPoderoso — CHANGES_REQUESTED** [review]

File: src/cpp/rtps/builtin/discovery/participant/PDPClient.cpp [hunk 4]
```diff
@@ -769,9 +769,17 @@ void PDPClient::announceParticipantState(
...
                         {
+                            GuidPrefix_t srv_guid_prefix = svr.guidPrefix;
+#if HAVE_SECURITY
+                            if (getRTPSParticipant()->is_secure())
+                            {
+                                // Need the mangled guid prefix in this case
+                                srv_guid_prefix = get_participant_proxy_data(svr.guidPrefix)->m_guid.guidPrefix;
+                            }
+#endif  // HAVE_SECURITY
                             //locators.push_back(svr.metatrafficMulticastLocatorList);
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/discovery/participant/PDPClient.cpp:780]:
```suggestion
#endif  // HAVE_SECURITY

```

**Review by MiguelCompany — CHANGES_REQUESTED** [review]
Just a NIT

**Review by MiguelCompany — APPROVED** [review]

**Pull request merged**